### ISSUE TITLE
Percentage based rules

### ISSUE THREAD

**Issue by AJenbo** [phpmd/phpmd#895]
## Description
[ISO/IEC 25010:2011](https://www.iso.org/standard/35733.html) is an industry standard for code maintainability, it is often set as a sales requirement for software that it must live up to this standard.

There are tools that will check the status of a project as well as CI services, but there does not appear to be any developer tools that would help to verify the adherence as the developer works on the code.

PHPMD can analyze and catch a lot of the patterns described in the standard. But one aspect that PHPMD is not able to reflect is the allowance of gradation in code that falls outside the idea. For example PHPMD will simply report any function that is longer then a configured max of 15 lines, but ISO-25010 allows for 44% of the code to be in functions of 15 or more lines, 22% can be 30+ and 7% maybe be 60+. Only when one of these limits are broken should the functions in that range be reported ass violations.

In order to facilitate this feature we would need to expand the rule set to allow multiple entries for the same rule.
The analyzer would also need to know the total line number before it concludes what is a violation.

To keep backwards comparability any rule that does not have a percentage defined should assume that 100% of the code must comply with the rule.

Suggested config format:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset>
    <rule ref="rulesets/codesize.xml/ExcessiveMethodLength">
        <properties>
            <property name="minimum" value="15" precentage="56"/>
            <property name="ignore-whitespace" value="true"/>
        </properties>
        <properties>
            <property name="minimum" value="31" precentage="78"/>
            <property name="ignore-whitespace" value="true"/>
        </properties>
        <properties>
            <property name="minimum" value="60" precentage="93"/>
            <property name="ignore-whitespace" value="true"/>
        </properties>
    </rule>
</ruleset>
```

**Comment by ravage84** [comment]
Interesting idea.

I question whether this should be the responsibility of PHPMD itself, though. May be I'm just coming too much from a "[do one thing and do it well](https://en.wikipedia.org/wiki/Unix_philosophy)" point of view....
